### DISCUSSION TITLE
How to import open tabs previously exported from rooted device

### DISCUSSION THREAD

**Discussion by cmonty14** [bromite/bromite#2589] [Q&A]
Hello,
I have exported anything incl. open tabs from rooted device based on [this](https://github.com/bromite/bromite/wiki/Migrating-user-data) article from 2018.

How can I import open tabs on a non-rooted device?
Where is all data stored after installing Bromite on non-rooted device?

THX

**Comment by SebiderSushi** [comment]
Data is stored in the same location regardless of root status.

 - If you are using a `userdebug` build you can enable rooted ADB debugging and do the file operations in a root shell: <code>adb&nbsp;root&nbsp;&&&nbsp;adb&nbsp;shell</code>  
   You can check your build type with `adb shell getprop ro.build.type`. A `user` build won't allow rooted ADB. If your build type is `eng` you should not use that ROM with any sensitive data, `eng` build are only meant for debugging.
 - If you have a custom recovery like TWRP you can do the file operations in recovery, either on the device or via <code>adb&nbsp;shell</code>

This should work:
1. Install Bromite on the device
2. Launch it at least once to let it create all of its files
3. Exit Bromite and close the app completely or reboot to recovery.
4. Open a rooted shell via ADB or use the TWRP terminal or file manager
5. Follow the guide [you linked](https://github.com/bromite/bromite/wiki/Migrating-user-data) and restore the files you want.  
   - If you're restoring single files and Bromite already created them you can use `cat [backup file] > [target file]`. This will keep the existing file ownership, SELinux context and permissions so you can skip the next step for these files.
7. After restoring you need to set the correct file permissions:
    - Run `ls -Zl`, note down owner, group, and SELinux context from one file or folder you did not touch.
    - Run `chown -R u0_app123:u0_app123 [path]` with the correct owner:group on any file or folder you restored.
    - Run `chcon -R u:object_r:app_data_file:s0:c181,c256,c512,c768 [path]` with the correct context on any file or folder you restored.
8. Launch Bromite and check if everything works as expected.
